libfreerdp/core/license.c

```diff
@@ -1703,7 +2646,173 @@ void license_free(rdpLicense* license)
+			return "LICENSE_STATE_CONFIGURED";
+		case LICENSE_STATE_REQUEST:
+			return "LICENSE_STATE_REQUEST";
+		case LICENSE_STATE_NEW_REQUEST:
+			return "LICENSE_STATE_NEW_REQUEST";
+		case LICENSE_STATE_PLATFORM_CHALLENGE:
+			return "LICENSE_STATE_PLATFORM_CHALLENGE";
+		case LICENSE_STATE_PLATFORM_CHALLENGE_RESPONSE:
+			return "LICENSE_STATE_PLATFORM_CHALLENGE_RESPONSE";
+		case LICENSE_STATE_COMPLETED:
+			return "LICENSE_STATE_COMPLETED";
+		case LICENSE_STATE_ABORTED:
+			return "LICENSE_STATE_ABORTED";
+		default:
+			return "LICENSE_STATE_UNKNOWN";
+	}
+}
+
+BOOL license_server_send_request(rdpLicense* license)
+{
+	WINPR_ASSERT(license);
+	if (!license_ensure_state(license, LICENSE_STATE_CONFIGURED, LICENSE_REQUEST))
+		return FALSE;
+	if (!license_send_license_request_packet(license))
+		return FALSE;
+	return license_set_state(license, LICENSE_STATE_REQUEST);
+}
+
+BOOL license_server_configure(rdpLicense* license)
+{
+
+	int len;
+	wStream* s;
+	UINT32 algs[] = { KEY_EXCHANGE_ALG_RSA };
+	UINT32 x;
+	const rdpSettings* settings;
+	UINT32 ProductVersion, issuerCount;
+	const char* CompanyName;
+	const char* ProductName;
+	const char** issuers;
+
+	WINPR_ASSERT(license);
+	WINPR_ASSERT(license->rdp);
+
+	settings = license->rdp->settings;
+	WINPR_ASSERT(settings);
+
+	CompanyName = freerdp_settings_get_string(settings, FreeRDP_ServerLicenseCompanyName);
+
+	ProductName = freerdp_settings_get_string(settings, FreeRDP_ServerLicenseProductName);
+	ProductVersion = freerdp_settings_get_uint32(settings, FreeRDP_ServerLicenseProductVersion);
+	issuerCount = freerdp_settings_get_uint32(settings, FreeRDP_ServerLicenseProductIssuersCount);
+	issuers =
+	    (const char**)freerdp_settings_get_pointer(settings, FreeRDP_ServerLicenseProductIssuers);
+
+	WINPR_ASSERT(CompanyName);
+	WINPR_ASSERT(ProductName);
+	WINPR_ASSERT(ProductVersion > 0);
+	WINPR_ASSERT(issuers || (issuerCount == 0));
+
+	if (!license_ensure_state(license, LICENSE_STATE_INITIAL, LICENSE_REQUEST))
+		return FALSE;
+
+	license->ProductInfo->dwVersion = ProductVersion;
+	len = ConvertToUnicode(CP_UTF8, 0, CompanyName, -1,
+	                       (WCHAR**)&license->ProductInfo->pbCompanyName, 0);
+	if (!license->ProductInfo->pbCompanyName)
+		return FALSE;
+	license->ProductInfo->cbCompanyName = len * sizeof(WCHAR);
+
+	len = ConvertToUnicode(CP_UTF8, 0, ProductName, -1, (WCHAR**)&license->ProductInfo->pbProductId,
+	                       0);
+	if (!license->ProductInfo->pbProductId)
+		return FALSE;
+	license->ProductInfo->cbProductId = len * sizeof(WCHAR);
+
+	if (!license_read_binary_blob_data(license->KeyExchangeList, BB_KEY_EXCHG_ALG_BLOB, algs,
+	                                   sizeof(algs)))
+		return FALSE;
+
+	if (!certificate_read_server_certificate(license->certificate, settings->ServerCertificate,
+	                                         settings->ServerCertificateLength))
+		return FALSE;
+
+	s = Stream_New(NULL, 1024);
+	if (!s)
+		return FALSE;
+	else
+	{
```

Comment:
we can probably remove the `else` (not that we have upgraded C standard, declaring `r` right after should be ok)

Reply:
true, but I think the scope makes it more readable, but can change that.